tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc

```diff
@@ -525,39 +662,30 @@ StatusOr<bool> HorizontalLoopFusionImpl::Run() {
   // shape mismatch but bitcasts could prevent future h-fusion from happening.
   // So, a bottom-up, use-to-def order should be more favorable. It also helps
   // to save compiler iterations to reach the fixed point.
-  std::vector<HloInstruction*> use_to_def_order =
-      computation_->MakeInstructionPostOrder();
-  absl::c_reverse(use_to_def_order);
-  for (size_t i = 0; i < use_to_def_order.size(); ++i) {
-    HloInstruction* consumer = use_to_def_order[i];
-    HorizontalLoopFusionImpl::FusionCandidates fusion_candidates(consumer);
-    while (true) {
-      auto fusibles = fusion_candidates.GetNextSpanOfFusions();
-      if (fusibles.empty()) {
-        break;
-      } else if (fusibles.size() == 1) {
-        // Skip; there is just one fused_instr.
-        continue;
-      }
+  std::vector<HloInstruction*> to_fuse_candidates = computation_->MakeInstructionPostOrder();
 
-      changed = true;
-      // Convert fusible into fusion_instrs to simplify the implementation of
-      // `Fuse()`.
-      std::vector<HloInstruction*> fusion_instrs;
-      for (HloInstruction* instr : fusibles) {
-        if (instr->opcode() == HloOpcode::kFusion) {
-          fusion_instrs.push_back(instr);
-        } else {
-          TF_ASSIGN_OR_RETURN(
-              HloInstruction * fusion_instr,
-              MakeFusionInstruction(instr, HloInstruction::FusionKind::kLoop));
-          fusion_instrs.push_back(fusion_instr);
-        }
-      }
-      TF_RETURN_IF_ERROR(Fuse(absl::MakeSpan(fusion_instrs)));
+  while (to_fuse_candidates.size() > 0 ) {
+    HloInstruction* consumer = to_fuse_candidates.back();
+    to_fuse_candidates.pop_back();
+
+    // the consumer may be the operands of previously fused instruction, so 
+    // it will no longer valid, skip this instruction.
+    if (consumer->IsDead()) {
```

Comment:
If consumer is dead, shouldn't it be removed? This may also fix the issue that Reed reported.

Reply:
Thanks @akuegel for the notes, sorry that the response is slow, I am OOTO this week, will work the fix on return.

Reply:
@reedwm @akuegel  I submitted a fix for the unittest failure. 
Seems that we should run DCE pass after horizontal fusion optimization, launched [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/gpu_compiler.cc#L823) .

Reply:
Sure, that is a solution, but it is kind of a big hammer. Previously, we had made this pass delete the dead ops it created itself. Did you also try to just delete the consumer in line 673 if IsDead() returns true, and that still did not fix the problem? In the case it is not clear where the dead ops come from, I am fine with adding DCE after HorizontalFusion.

Reply:
Yes, I tried.  The senario when just removing the dead instruction is little bit complex due to the process order of candidate instructions.  There is a case when the instruction is dead, but it has already removed from the computation.  So the simplest fix might be just calling DCE after horizontal fusion. 

I checked the DCE pass, since the implementation is realitively simple, so just calling extra time looks ok to me.   I can try to do the fix in horizontal fusion if this fix shows to be an issue.

Reply:
Unfortunately it turns out, adding HloDCE pass does not fix the issue. The GPU-ir-emit-prepare pipeline is configured to run the HloVerifier after each pass, so it will run after GpuHorizontalLoopFusion pass before HloDCE did run. Can you please try to figure out how to delete dead instructions as part of GpuHorizontalLoopFusion?

Reply:
Also, just to make it clear: this PR was merged, but then automatically rolled back because it made a test fail when run in debug mode. I don't know of an easy way to fix this. Can you please recreate this PR with a fix?

Reply:
Ok, I actually found a solution: We can create a sub-pipeline that contains both the HorizontalFusion and the DCE pass, and then the verifier will only be run after both have run, and then it is ok. I will reland your PR with that fix.